python/cudf/cudf/utils/utils.py

```diff
@@ -400,3 +402,14 @@ def _maybe_indices_to_slice(indices: cp.ndarray) -> Union[slice, cp.ndarray]:
     if (indices == cp.arange(start, stop, step)).all():
         return slice(start, stop, step)
     return indices
+
+
+def cudf_nvtx_annotate(func, color="rapids", domain="cudf_python"):
+    return annotate(message=func.__qualname__, color=color, domain=domain)(
+        func
+    )
```

Comment:
What about doing something clever with using the message contents to define a unique color so different function ranges have different colors?

Reply:
Part of why I suggested defining this wrapper is so that we could centralize that kind of think, it definitely could be helpful. What kinds of groupings do you think are most important, though? For instance, we could use `func.__module__`, but that may not be the most useful differentiator.

Reply:
That is possible, any preferences on what colors should be assigned to which type of APIs?

cc: @shwina

Reply:
No idea. I might just start with a different color per function name and see where that gets us.

Reply:
Since we don't know a priori how many colors that we'll need, to do that we'd need to generate them on the fly. The simplest implementation I can think of is something like this:
```
_cur_color = 0

def _make_color():
    global _cur_color
    # Cycle through colors. I randomly picked a random prime number < 1000.
    _cur_color = (_cur_color + 997) / int(0xffffff)
    return _cur_color
```
Then in `cudf_nvtx_annotate` you can do `annotate(..., color=make_color(), ...)` and get a unique color every time. I can't think of an easy way that isn't a ton of work to generate a lookup table with exactly the right number of colors ahead of time, though. If you really wanted, it would be straightforward to add some logic to store the colors in a global lookup table in the package that you could check after the fact by importing.

Reply:
I would suggest keeping things simple here and drawing from a small set of colors that are easy on the eyes. We can revisit the idea of generating unique colors per function if necessary but I haven't found the profiles generated so far difficult to interpret.

Reply:
Like @shwina said, We could do something like cycling between a few subset of colors based hash of `func. __qualname__`:

```python
nvtx_colors = ["green", "blue", "purple", "rapids"]
def get_color(name):
    idx = hash(name) % 4
    return nvtx_colors[idx]

annotate(message=func.__qualname__, color=get_color(func.__qualname__), ...)
```

Reply:
As a reminder, the builtin `hash` is not reproducible across multiple processes or multiple invocations. If we want the colors to be consistent across multiple runs (or dask workers?), we would need to use a deterministic hash.

> By default, the [`__hash__()`](https://docs.python.org/3.4/reference/datamodel.html#object.__hash__) values of str, bytes and datetime objects are “salted” with an unpredictable random value. Although they remain constant within an individual Python process, they are not predictable between repeated invocations of Python.

https://docs.python.org/3.4/reference/datamodel.html#object.__hash__

Potential replacement:

```python
import hashlib

nvtx_colors = ["green", "blue", "purple", "rapids"]
def get_color(name):
    m = hashlib.sha256()
    m.update(name.encode())
    hash_value = int(m.hexdigest(), 16)
    idx = hash_value % len(nvtx_colors)
    return nvtx_colors[idx]

annotate(message=func.__qualname__, color=get_color(func.__qualname__), ...)